resources/api/resource.py

```diff
@@ -343,28 +381,72 @@ def _is_resource_open(self, resource, start, end):
 
         return True
 
-    def filter_available_between(self, queryset, name, value):
-        if len(value) != 2:
-            raise exceptions.ParseError('available_between takes exactly two comma-separated values.')
+    def _filter_available_between_with_period(self, queryset, reservations, available_start, available_end, period):
+        reservations = reservations.order_by('begin').select_related('resource')
 
-        available_start = self._deserialize_datetime(value[0])
-        available_end = self._deserialize_datetime(value[1])
+        reservations_by_resource = collections.defaultdict(list)
+        for reservation in reservations:
+            reservations_by_resource[reservation.resource_id].append(reservation)
 
-        if available_start.date() != available_end.date():
-            raise exceptions.ParseError('available_between timestamps must be on the same day.')
+        available_resources = set()
 
-        # exclude resources that have reservation(s) overlapping with the available_between range
-        overlapping_reservations = Reservation.objects.filter(
-            resource__in=queryset, end__gt=available_start, begin__lt=available_end
-        ).current()
-        queryset = queryset.exclude(reservations__in=overlapping_reservations)
-        closed_resource_ids = {
-            resource.id
-            for resource in queryset
-            if not self._is_resource_open(resource, available_start, available_end)
-        }
+        hours_qs = ResourceDailyOpeningHours.objects.filter(
+            open_between__overlap=(available_start, available_end, '[)'))
 
-        return queryset.exclude(id__in=closed_resource_ids)
+        # check the resources one by one to determine which ones have open slots
+        for resource in queryset.prefetch_related(None).prefetch_related(
+                Prefetch('opening_hours', queryset=hours_qs, to_attr='prefetched_opening_hours')):
+            reservations = reservations_by_resource[resource.id]
+
+            if self._is_resource_available(resource, available_start, available_end, reservations, period):
+                available_resources.add(resource.id)
+
+        return queryset.filter(id__in=available_resources)
+
+    @staticmethod
+    def _is_resource_available(resource, available_start, available_end, reservations, period):
+        opening_hours = resource.get_opening_hours(available_start, available_end, resource.prefetched_opening_hours)
+        hours = next(iter(opening_hours.values()))[0]  # assume there is only one hours obj per day
```

Comment:
Is this true enough? I've heard rumblings of places with self-service hours and full-service hours… No. I've no idea if they are affect this level of abstraction. 😬